pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,971 @@
+					}
+					Ok(())
+				},
+			)?;
+
+			Self::remove_schedule_from_storages(&schedule.owner, schedule_id);
+
+			Self::deposit_event(Event::Terminated {
+				id: schedule_id,
+				who: schedule.owner,
+				error: Error::<T>::ManuallyTerminated.into(),
+			});
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T>
+where
+	<T as pallet_route_executor::Config>::Balance: From<Balance>,
+	Balance: From<<T as pallet_route_executor::Config>::Balance>,
+{
+	fn prepare_schedule(
+		current_blocknumber: T::BlockNumber,
+		weight_for_dca_execution: Weight,
+		schedule_id: ScheduleId,
+		schedule: &Schedule<T::AccountId, T::AssetId, T::BlockNumber>,
+	) -> DispatchResult {
+		Self::take_transaction_fee_from_user(schedule_id, schedule, weight_for_dca_execution)?;
+
+		if Self::price_change_is_bigger_than_max_allowed(schedule) {
+			Self::retry_schedule(schedule_id, schedule, current_blocknumber)?;
+
+			return Err(Error::<T>::PriceChangeIsBiggerThanMaxAllowed.into());
+		}
+
+		Ok(())
+	}
+
+	#[transactional]
+	pub fn execute_trade(
+		schedule_id: ScheduleId,
+		schedule: &Schedule<T::AccountId, T::AssetId, T::BlockNumber>,
+	) -> DispatchResult {
+		let origin: OriginFor<T> = Origin::<T>::Signed(schedule.owner.clone()).into();
+
+		match &schedule.order {
+			Order::Sell {
+				asset_in,
+				asset_out,
+				amount_in,
+				min_limit,
+				slippage,
+				route,
+			} => {
+				let remaining_amount_to_use =
+					RemainingAmounts::<T>::get(schedule_id).defensive_ok_or(Error::<T>::InvalidState)?;
+				let amount_to_sell = min(remaining_amount_to_use, *amount_in);
+
+				Self::unallocate_amount(schedule_id, schedule, amount_to_sell)?;
+
+				let (estimated_amount_out, slippage_amount) =
+					Self::calculate_estimated_and_slippage_amounts(*asset_out, *asset_in, amount_to_sell, *slippage)?;
+				let min_limit_with_slippage = estimated_amount_out
+					.checked_sub(slippage_amount)
+					.ok_or(ArithmeticError::Overflow)?;
+				let min_limit = max(*min_limit, min_limit_with_slippage);
+
+				let route = route.to_vec();
+				let trade_amounts =
+					pallet_route_executor::Pallet::<T>::calculate_sell_trade_amounts(&route, amount_to_sell.into())?;
+				let last_trade = trade_amounts.last().defensive_ok_or(Error::<T>::InvalidState)?;
+				let amount_out = last_trade.amount_out;
+
+				ensure!(amount_out >= min_limit.into(), Error::<T>::TradeLimitReached);
+
+				pallet_route_executor::Pallet::<T>::sell(
+					origin,
+					*asset_in,
+					*asset_out,
+					(amount_to_sell).into(),
+					min_limit.into(),
+					route,
+				)
+			}
+			Order::Buy {
+				asset_in,
+				asset_out,
+				amount_out,
+				slippage,
+				max_limit,
+				route,
+			} => {
+				let amount_in = Self::get_amount_in_for_buy(amount_out, route)?;
+
+				Self::unallocate_amount(schedule_id, schedule, amount_in)?;
+
+				let (estimated_amount_in, slippage_amount) =
+					Self::calculate_estimated_and_slippage_amounts(*asset_in, *asset_out, *amount_out, *slippage)?;
```

Comment:
`calculate_dynamic_slippage` would be probably more accurate name

Reply:
i was confused why are we still "estimating" amount when we have them exactly calculated and then i recalled that we are calculating another limit based on the last block price to prevet frontrunning :)